### PR TITLE
Store measurement metadata

### LINKED ISSUE #413: Enhance nodes API to accept measurements generated by Project Tango
As requested by SMF, the nodes API should be enhanced in a way so measurements can be stored for each node.

Pasting the proposal from SMF here:

```
Messungsbild hochladen:

/api/nodes/{node_id}/measurement
POST (wie /api/nodes/{node_id}/photos)

Response:
{"message":"OK", "id": 123456}

===========================================

Metadaten hochladen:

/api/nodes/{node_id}/measurement/{id}/meta
PUT oder POST

{
        'type': 'DOOR/STEPS/RAMP/TOILET',
        'description': 'some user description',
        'data': {
                // daten abhängig von dem typ
                // DOOR
                'width': 1.04 //meter

                // STEPS
                'height': 0.11 //meter

                // RAMP
                'angle': 15.42 // degree

                // TOILET
                'space': 2.04 // m^2

        }
}
```

A measurement consists for an image (Messungsbild) and meta data such as:
- width (in meters)
- height (in meters)
- angle (in degrees)
- space (in m^2)

### PR DESCRIPTION
**Pull request by schultyy**
Resolves #413 

This adds the implementation for storing measurement metadata. Metadata can be attached to any photo specified. Depending on the metadata type (`ramp`, `toilet`, `step`, `door`) different types of parameters are allowed. See https://github.com/sozialhelden/wheelmap/issues/413#issuecomment-251432176 for details.

Usage example:

```
$ curl -v -XPOST -H "Content-Type: application/json" -H "Accept: application/json" "https://staging.wheelmap.org/api/nodes/371857052/measurements/23611/metadata?api_key=<API KEY>" -d '{"metadata": {"data_type": "door", "description": "Some description", "data": { "width": 15.0 }}}'
```

### CODE REVIEW

File: app/controllers/api/measurements_controller.rb [hunk 1]
```diff
@@ -1,6 +1,33 @@
...
+  rescue_from ActionController::UnpermittedValue do |e|
+    respond_to do |wants|
+      wants.json{ render :json => {:error => "Param #{e.key} must be one of the following values: #{e.allowed_values.inspect}"}.to_json, :status => 406 }
+      wants.xml{  render :xml  => {:error => "Param #{e.key} must be one of the following values: #{e.allowed_values.inspect}"}.to_xml,  :status => 406 }
+    end
+  end
+
+  resource_description do
+    short 'Measurements. Provided by Google Project Tango enabled devices'
+    error :code => 401, :desc => "Authorization Required", meta: { message: "Authentication failed or was not provided. Verify that you have sent valid credentials via an api_key parameter. A 'Www-Authenticate' challenge header will be sent with this type of error response." }
```
Comment (Hoverbear) [inline app/controllers/api/measurements_controller.rb]: `Www-Authenticate` -> `www-Authenticate`

File: app/controllers/api/measurements_controller.rb [hunk 2]
```diff
@@ -1,6 +1,33 @@
...
+    formats ['json']
+    param :api_key, String, desc: "You personal API key. Sign up for an account at http://wheelmap.org/users/sign_in", required: true
```
Comment (Hoverbear) [inline app/controllers/api/measurements_controller.rb]: `You` -> `Your`

File: app/controllers/api/measurements_controller.rb [hunk 3]
```diff
@@ -1,6 +1,33 @@
...
+
+  def_param_group :data do
+    param :height, Float
+    param :angle, Float
+    param :area, Float
+    param :width, Float
+  end
+
+  def_param_group :measurement do
+    param :type, String, required: true, desc: "The measurement's type. Can be door, toilet, step or ramp"
```
Comment (Hoverbear) [inline app/controllers/api/measurements_controller.rb]: `"The measurement's type. Can be door, toilet, step or ramp"` -> `"The measurement type. Can be: 'door', 'toilet', 'step', or 'ramp'"`

File: app/models/measurement.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    when 'ramp'
+      validate_ramp_point
+    end
+  end
+
+  private
+
+  def validate_door_point
+    return if datapoints.size == 0
+    errors.add(:datapoints, "door must have a width") if datapoints.first.property != 'width'
```
Comment (Hoverbear) [inline app/models/measurement.rb]: So only the first `datapoint` of a `Measurement` is validated?

File: spec/controllers/api/measurements_controller_spec.rb [hunk 5]
```diff
@@ -107,77 +150,363 @@ def json_response
...
+        it 'accepts valid json' do
+          expect(response.status).to eq 201
+        end
+
+        describe 'metadata' do
+          let(:measurement) { picture.measurements.first }
+          let(:data_point) { measurement.datapoints.first }
+
+          it 'stores exactly one data point' do
+            expect(picture.measurements.length).to eq 1
```
Comment (Hoverbear) [inline spec/controllers/api/measurements_controller_spec.rb]: This is checking `measurements.length` not `measurements.datapoints.length` like the test description suggests.

File: spec/controllers/api/measurements_controller_spec.rb [hunk 6]
```diff
@@ -107,77 +150,363 @@ def json_response
...
+          post(:add_metadata, :node_id => poi.id, :measurement_id => picture.id, :api_key => user.authentication_token, :metadata => valid_steps_metadata)
+          expect(response.status).to eq 201
+        end
+
+        describe "metadata" do
+          let(:measurement) { picture.measurements.first }
+          let(:data_point) { measurement.datapoints.first }
+
+          it 'stores exactly one data point' do
+            expect(picture.measurements.length).to eq 1
```
Comment (Hoverbear) [inline spec/controllers/api/measurements_controller_spec.rb]: Similar as above, this is checking `measurements` not `measurements.datapoints` as the description suggests.

File: spec/controllers/api/measurements_controller_spec.rb [hunk 7]
```diff
@@ -107,77 +150,363 @@ def json_response
...
+        it 'accepts valid json' do
+          expect(response.status).to eq 201
+        end
+
+        describe "metadata" do
+          let(:measurement) { picture.measurements.first }
+          let(:data_point) { measurement.datapoints.first }
+
+          it 'stores exactly one data point' do
+            expect(picture.measurements.length).to eq 1
```
Comment (Hoverbear) [inline spec/controllers/api/measurements_controller_spec.rb]: Similar as above, this is checking `measurements` not `measurements.datapoints` as the description suggests.

File: app/models/measurement.rb (hunk 4 above)
Comment (schultyy) [inline app/models/measurement.rb]: Yes, all measurements have just one data point except toilet. Then it needs to have exactly three: 
- width
- length
- area

Comment (Hoverbear) [inline app/models/measurement.rb]: Okay :)

**Comment by Hoverbear** [comment]
`measurement_type` is reserved? Huh.

**Comment by schultyy** [comment]
As requested by Tim Freiheit, it is now possible to save metadata with an empty description.

**Pull request merged**